kanren/cons.py

```diff
@@ -0,0 +1,246 @@
+from functools import reduce
+
+from itertools import chain
+
+from collections import OrderedDict
+
+try:
+    from operator import length_hint
+except ImportError:
+    def length_hint(obj, default=0):
+        """Return an estimate of the number of items in obj.
+        From https://www.python.org/dev/peps/pep-0424/
+        """
+        try:
+            return len(obj)
+        except TypeError:
+            try:
+                get_hint = type(obj).__length_hint__
+            except AttributeError:
+                return default
+            try:
+                hint = get_hint(obj)
+            except TypeError:
+                return default
+            if hint is NotImplemented:
+                return default
+            if not isinstance(hint, int):
+                raise TypeError("Length hint must be an integer, not %r" %
+                                type(hint))
+            if hint < 0:
+                raise ValueError("__length_hint__() should return >= 0")
+            return hint
+
+try:
+    from collections.abc import Iterator
+except ImportError:
+    from collections import Iterator
+
+from multipledispatch import dispatch
+
+from toolz import drop, last
+
+from six import add_metaclass
+
+from unification.core import unify, _unify, reify, _reify
+
+
+def first(x):
+    return next(iter(x), None)
+
+
+def rest(x):
+    return drop(1, x)
+
+
+def none_to_list(x):
+    return [] if x is None else x
+
+
+@dispatch(object, object)
+def cons_merge(car_part, cdr_part):
+    """Merge a generic car and cdr."""
+    return None
+
+
+class ConsType(type):
+    def __instancecheck__(self, o):
+        if is_cons(o):
+            return True
+        return False
+
+
+@add_metaclass(ConsType)
+class ConsPair(object):
+    """An object representing cons pairs.
+
+    These objects, and the class constructor alias `cons`, serve as a sort of
+    generalized delayed append operation for various collection types.  When
+    used with the built-in Python collection types, `cons` behaves like the
+    concatenate operator between the given types, if any.
+
+    A Python `list` is returned when the cdr is a `list` or `None`; otherwise,
+    a `ConsPair` is returned.
+
+    The arguments to `ConsPair` can be a car & cdr pair, or a sequence of
+    objects to be nested in `cons`es, e.g.
+
+        ConsPair(car_1, car_2, car_3, cdr) ==
+            ConsPair(car_1, ConsPair(car_2, ConsPair(car_3, cdr)))
+    """
+
+    def __new__(cls, *parts):
```

Comment:
Maybe add `__slots__` or inherit from `tuple` for memory efficiency?

Reply:
Inheriting from `tuple` would break the consistency with Lisp `cons`, since `tuple`s are necessarily iterable and a `cons`-pair isn't.

I added `__slots__`, though.